### PR TITLE
feat: add ckb mbt proof verify precompile contract

### PR DESCRIPTION
**Pull request by KaoImin**
## What this PR does / why we need it?

This PR add [ckb-mbt](https://docs.rs/ckb-types/0.112.0/ckb_types/utilities/type.CBMT.html)(aka merkle binary tree) [proof](https://docs.rs/ckb-types/0.112.0/ckb_types/utilities/type.MerkleProof.html) verify precompile contract. The input struct is
```rust
#[derive(EthAbiCodec, EthAbiType, Clone, Debug, PartialEq, Eq)]
pub struct VerifyProofPayload {
    /// If the verify_type is 0, the leaves should be in the
    /// raw_transactions_root, otherwise in the witnesses_root.
    pub verify_type:           u8,
    pub transactions_root:     [u8; 32],
    pub witnesses_root:        [u8; 32],
    pub raw_transactions_root: [u8; 32],
    pub proof:                 Proof,
}

#[derive(EthAbiCodec, EthAbiType, Clone, Debug, PartialEq, Eq)]
pub struct Proof {
    pub indices: Vec<u32>,
    pub lemmas:  Vec<[u8; 32]>,
    pub leaves:  Vec<[u8; 32]>,
}
```
Because of [the transactions_root in a CKB block header](https://github.com/nervosnetwork/rfcs/blob/01f3bc64ef8f54c94c7b0dcf9d30c84b6c8418b0/rfcs/0027-block-structure/0027-block-structure.md#transactions_root-byte32) is consist of two layers:
```
                                                                          
                                  Transactions Root                        
                                        /   \                              
                                       /     \                             
                                      /       \                            
                                     /         \                           
                                    v           v                          
                  Raw Transactions Root          Witnesses Root            
                    /       |     \               |         \              
                   /        |      \              |          \             
                  /         |       \             |           \            
                 v          v        v            v            v           
            TxHash00    TxHash01   TxHash02   WitnessHash00   WitnessHash01
                                                                           
```
So apart from the transactions, witnesses root also needs. And it should be use ABI encode to bytes such as:
```solidity
abi.encode(VerifyProofPayload(transactions_root, witnesses_root, raw_transactions_root, indices, lemmas, leaves))
```

### What is the impact of this PR?

No Breaking Change

**PR relation**:
- Ref #

<details><summary>CI Settings</summary><br/>

### **CI Usage**

**Tip**: Check the CI you want to run below, and then comment `/run-ci`.

**CI Switch**

- [ ] Web3 Compatible Tests
- [ ] OCT 1-5 And 12-15
- [ ] OCT 6-10
- [ ] OCT 11
- [ ] OCT 16-19
- [ ] v3 Core Tests

### **CI Description**

| CI Name                                   | Description                                                               |
| ----------------------------------------- | ------------------------------------------------------------------------- |
| *Web3 Compatible Test*                    | Test the Web3 compatibility of Axon                                       |
| *v3 Core Test*                            | Run the compatibility tests provided by Uniswap V3                        |
| *OCT 1-5 \| 6-10 \| 11 \| 12-15 \| 16-19* | Run the compatibility tests provided by OpenZeppelin                      |

</details>

### CODE REVIEW

**Comment by jjyr** [comment]
This contract only supports verify raw_transaction, but users may want to also verify the witness transaction(to ensure the witness part is valid on CKB). This contract must support merkle verify the `witness_root`.

I suggest to add another input field `tx_verify_type: u8` to indicates wether we want to verify a raw_transaction or a witness_transaction.

If the `tx_verify_type` is `0`, the merkle root must equals `raw_transaction_root`, if the `tx_verify_type` is `1` the merkle root must equals to `witness_transaction_root`.

File: core/executor/src/precompiles/ckb_mbt_verify.rs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+#[derive(EthAbiCodec, EthAbiType, Clone, Debug, PartialEq, Eq)]
+pub struct VerifyProofPayload {
+    pub transactions_root:     [u8; 32],
+    pub witnesses_root:        [u8; 32],
+    pub raw_transactions_root: [u8; 32],
+    pub indices:               Vec<u32>,
+    pub lemmas:                Vec<[u8; 32]>,
+    pub leaves:                Vec<[u8; 32]>,
+}
```
Comment (ashuralyk) [inline core/executor/src/precompiles/ckb_mbt_verify.rs]: how can I get this payload via CKB-RPC?

Comment (KaoImin) [reply core/executor/src/precompiles/ckb_mbt_verify.rs]: The `witnesses_root`, `indices`, `lemmas` use [this](https://github.com/nervosnetwork/ckb/blob/develop/rpc/README.md#method-get_transaction_proof) RPC.
The `leaves` should be **all** of the transaction hashes in this block.

Comment (ashuralyk) [reply core/executor/src/precompiles/ckb_mbt_verify.rs]: how can I verify a cell belongs to a transaction? and to verify the transaction itself, I can verify its hash is included in the `transacitons_root`?

**Comment by KaoImin** [comment]
> This contract only supports verify raw\_transaction, but users may want to also verify the witness transaction(to ensure the witness part is valid on CKB). This contract must support merkle verify the `witness_root`.
> 
> I suggest to add another input field `tx_verify_type: u8` to indicates wether we want to verify a raw\_transaction or a witness\_transaction.
> 
> If the `tx_verify_type` is `0`, the merkle root must equals `raw_transaction_root`, if the `tx_verify_type` is `1` the merkle root must equals to `witness_transaction_root`.

Sounds good.

**Review by jjyr — APPROVED** [review]

File: core/executor/src/precompiles/tests.rs [hunk 2]
```diff
@@ -203,3 +214,100 @@ fn test_blake2f() {
...
+        set.into_iter().collect()
+    };
+
+    // Tx order is 0, 1, 2
+    let tx_hash_0 = H256::from_low_u64_be(0x00).0.pack();
+    let tx_hash_1 = H256::from_low_u64_be(0x01).0.pack();
+    let tx_hash_2 = H256::from_low_u64_be(0x02).0.pack();
+
+    let witness_hash_0 = H256::from_low_u64_be(0x10).0.pack();
+    let witness_hash_1 = H256::from_low_u64_be(0x11).0.pack();
```
Comment (ashuralyk) [inline core/executor/src/precompiles/tests.rs:234]: what is witness hash and how to get it?

Comment (KaoImin) [reply core/executor/src/precompiles/tests.rs:234]: Witnesses hashes is calculate as 

https://github.com/nervosnetwork/ckb/blob/378835b4eaebcd693d932ef6f0264b186c8c26c5/util/gen-types/src/extension/calc_hash.rs#L265-L271
https://github.com/nervosnetwork/ckb/blob/378835b4eaebcd693d932ef6f0264b186c8c26c5/util/gen-types/src/extension/calc_hash.rs#L144-L149

**Review by ashuralyk — APPROVED** [review]

**Pull request merged**